compiler/rustc_ast_lowering/src/delegation.rs

```diff
@@ -259,44 +260,105 @@ impl<'hir> LoweringContext<'_, 'hir> {
                         self_param_id: pat_node_id,
                     };
                     self_resolver.visit_block(block);
-                    let block = this.lower_block(block, false);
-                    this.mk_expr(hir::ExprKind::Block(block, None), block.span)
+                    let mut block = this.lower_block_noalloc(block, false);
+                    if block.expr.is_none() {
+                        let pat_hir_id = this.lower_node_id(pat_node_id);
+                        let arg = &*this.arena.alloc(this.generate_arg(pat_hir_id, span));
+                        // Use `self` as a first argument if no expression is provided.
+                        block.expr = Some(arg);
+                    }
+                    target_expr = Some(block);
                 } else {
                     let pat_hir_id = this.lower_node_id(pat_node_id);
-                    this.generate_arg(pat_hir_id, span)
+                    let arg = this.generate_arg(pat_hir_id, span);
+                    args.push(arg);
                 };
-                args.push(arg);
             }
 
-            let final_expr = this.finalize_body_lowering(delegation, args, span);
+            let final_expr = this.finalize_body_lowering(delegation, target_expr, args, span);
             (this.arena.alloc_from_iter(parameters), final_expr)
         })
     }
 
-    // Generates fully qualified call for the resulting body.
+    // Generates expression for the resulting body. If possible, `MethodCall` is used
+    // instead of fully qualified call for the self type coercion.
     fn finalize_body_lowering(
         &mut self,
         delegation: &Delegation,
-        args: Vec<hir::Expr<'hir>>,
+        target_expr: Option<hir::Block<'hir>>,
```

Comment:
```suggestion
        target_expr: hir::Expr<'hir>>,
```
Could you centralize the target expression extraction either in `lower_delegation_body` or in `finalize_body_lowering`?
Right now it's messy and it's not clear what is happening.